README
Nonpareil - a calculator microassembler and simulator
Copyright 1995, 2003, 2004 Eric L. Smith <eric@brouhaha.com>
$Id$

-----------------------------------------------------------------------------

News:

Release 0.46 of Nonpareil is an interim development release.

In this release, the HP-25, HP-31E, HP-33C, HP-35, HP-37E, HP-41CV, HP-45,
HP-55, and HP-80 are simulated.

The menu commands other than "Quit" and "About" are not implemented.

There are a few known bugs in the simulations:

* The HP-55 simulation does not correctly support user registers
  .0 through .9.

-----------------------------------------------------------------------------

Background:

Nonpareil is a microcode-level simulator for electronic calculators.
Currently it is able to simulate some of the HP Classic and Woodstock
series calculators.

The HP-35, the first of HP's classic series of handheld calculators,
was introduced in 1972 and described as the "Electronic Slide Rule".
The HP-35 was a non-programmable scientific, and was followed by the
HP-45 improved scientific calculator, the HP-65 scientific
programmable with card reader, the HP-55 scientific programmable with
stopwatch, and the HP-70 and HP-80 business calculators.  All of these
calculators used essentially the same processor, which was also used
in the HP 3380 Integrator, the HP 5830 and HP 5840 Gas Chromatographs,
and the HP 1722 Oscilliscope.

In the HP Classic series processor, unlike modern microprocessors,
various instructions were interpreted by various chips.  The core
processor consists of the C&T (control and timing), A&R (arithmetic
and registers) and the ROM or QUAD ROM chips.  Additions chips may be
present for data storage or program storage.  Since different
calculator models have different combinations of these chips, some
instructions are not available on all models.  For instance, the
"delayed select" instructions are probably not implemented in the
HP-35 and HP-80 which use single ROMs, and the "search for label"
instruction is only available in the HP-65.

The design of the processor and the instruction set are described in
detail in United States patents 3,863,060 and 4,001,569.

I've typed in the source code from the '569, '379, and '060 patents
which cover the HP-45, HP-55, and HP-80 respectively.  The listings in
the patents were poorly reproduced.  Peter Monta found and corrected
several errors.  The code in the patents might not be identical to
that in production units; I plan to eventually dump the ROMs from
actual calculators.

Peter Monta also *optically* dumped the contents of the HP-35 ROMs,
and disassembled the resulting code.  I've cleaned it up a bit and
added labels using the HP-45 source code as a guide.
        http://www.pmonta.com/calculators/hp-35/

In 1975 HP introduced their second generation handheld calculators,
known internally as the Woodstock series.  These use an improved
processor architecture described in the article "Inside the New Pocket
Calculators" in the November 1975 issue of Hewlett-Packard Journal.
An online copy of this article is available from the Museum of HP
Calculators:
        http://www.hpmuseum.org/journals/woodb.htm

In 1978, HP introduced their third generation handheld calculators,
known internally as the Spice series.  These use essentially the same
processor architecture as the Woodstock series, with the addition of
an instruction to checksum a 1K region of ROM, used as part of a
self-test.

In 1979, HP introduced their first expandable scientific handheld
calculator, the HP-41C (Coconut).  It was the first calculator to use
the Nut processor architecture, which was also used in the Voyager series
(HP-11C, HP-12C, etc.) introduced in 1981.

-----------------------------------------------------------------------------

Nonpareil simulator:

Nonpareil is a microcode-level simulator for the hardware of the HP
Classic and Woodstock series of calculators.  It should be capable of
simulating the following calculators:

  Classic series:    HP-35, HP-45, HP-55, HP-70, HP-80
  Woodstock series:  HP-21, HP-22, HP-25, HP-25C, HP-27, HP-29C
  Spice series:      HP-31E, HP-32E, HP-33E, HP-33C, HP-37E
  Coconut:           HP-41C, HP-41CV

Nonpareil uses the Gtk+ 2.2 toolkit and Glib 2.2.  It should be
portable to Windows and MacOS X, but has only been tested on Linux.

Nonpareil uses a single command line argument to specify the KML file.
If no argument is found, nonpareil will try to find a file using the
concatentation of the name under which it was invoked and ".kml".
This allows you to (for instance) link nonpareil to "hp45" and have it
automatically try to use "hp45.kml".

Nonpareil attempts to maintain a simulation speed of 3500
microinstructions per second, the nominal rate of the real
calculators, so that the HP-55 timer mode runs at roughly the correct
rate.  Other than the HP-55, the calculators used an LC oscillator
that was not very accurate.  On a computer that is heavily loaded, the
microinstruction rate will be reduced, so the HP-55 timer mode (and
the undocumented HP-45 timer mode) will run slow.  Don't blame me if
you try to use it and burn your eggs.

If you want to use the undocumented timer mode of the HP-45, there are
comments in hp45.kml explaining the change necessary to that file.  On
a real HP-45 this feature was accessed by pressing RCL then the "right
half of the ENTER key".  The ENTER key actually covered two switches,
but the key was shaped so that it only pressed the left switch, and
the ROM was written so the right switch would provide the ENTER
function as well as allowing use of the timer mode.  It was possible
to either put a shim on the key, or to create a a phantom keystroke by
simultaneously pressing CHS, 7, and 8.  Nonpareil doesn't support multiple
key presses at all, so the KML change makes the simulator's ENTER
button produce the keycode for the right switch.

-----------------------------------------------------------------------------

Uasm microassembler:

Uasm requires a single command line argument to specify the source
code file.

Uasm derives names for the object and listing files from the source
filename by replacing the extension (everything after the last dot,
typically "asm") with "obj" or "lst".  If the source filename doesn't
have an extension uasm adds one.  The object file format is one word
per line with the five digit octal address, a colon and the
instruction word in hexadecimal.

I personally prefer my assembly source code to be in all lower case,
so it doesn't look like it is shouting at me.  All the source code I
have typed in from patents is in lower case, but uasm is case
insensitive so if you prefer emphatic source code be my guest.

Uasm considers the keywords which are used to form instructions to be
reserved words, so they can't be used as labels.  HP's assembler was
more flexible in this regard; I had to change the label "go" in the
HP-55 source code to "go_".

Uasm does a better job of displaying the branch target addresses than
the HP assembler did.  It will actually show the correct target
address based on the delayed select instructions, unless the delay
instruction is in the last word of a page.  Uasm has a ".rom"
directive which can be used to provide the rom number.  This is
roughly equivalent to the origin directive in a typical assembler,
except that uasm always start assembling code at location 0 of the
selected ROM.

Because HP assembled each ROM separately, they were able to use the
same symbols in multiple ROMs.  uasm therefore has a separate name
space and keeps a separate symbol table for each ROM.

Uasm has a .symtab directive which causes the symbol table to be
printed to the listing file.  If the source file contains multiple
ROMs, each ROM should have its own .symtab directive.

The file "hp55.asm" when assembled with uasm will produce one warning,
which will be printed to standard out twice (once during each pass).
This warning can be ignored as it is the result of uasm being pickier
than the HP assembler about forms of the "go to" instruction.

-----------------------------------------------------------------------------

Development tools:

I used GNU tools to build nonpareil and uasm.  Originally I used GCC
2.5.8, Flex 2.4.7, Bison 1.22, and Make 3.71.  Most recently I'm using
GCC 3.3.2, Flex 2.5.4a, Bison 1.875, and Make 3.79.1.  Flex versions
prior to 2.4.2 will not work.

-----------------------------------------------------------------------------

Credits:

Nonpareil was originally written by Eric Smith.  Any bugs in it are
my fault.

Peter Monta provided the HP-35 code.  He also found and corrected
typographical errors in my entry of the HP-45 and HP-55 source code.

Images of the HP-25, HP-32E, HP-33C, HP-35, HP-37E, HP-45, HP-55, and
HP-80 are copyrighted by David G. Hicks of the Museum of HP
Calculators and have been included in this package by his permission:
        http://www.hpmuseum.org/

The HP-41C image (hp41cv.png) was developed by Maciej Barosiak.  Thomas
Olesen made improvements to hp41cv.kml for a much better looking display
and alpha keyboard mapping.

David G. Hicks translated the earlier csim HP-45 simulator to Java:
        http://www.hpmuseum.org/simulate/sim45.htm

Jonathan Purvis ported csim to PalmOS:
        http://one-two-three-four-five.com/palm/csim/

-----------------------------------------------------------------------------

Notices:

Nonpareil is not an HP product, and is not endorsed by HP.

Nonpareil is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License version 2 as
published by the Free Software Foundation.  Note that I am not
granting permission to redistribute or modify Nonpareil under the
terms of any later version of the General Public License.

Nonpareil is distributed in the hope that it will be useful (or at
least amusing), but WITHOUT ANY WARRANTY; without even the implied
warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
the GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program (in the file "COPYING"); if not, write to the
Free Software Foundation, Inc., 59 Temple Place - Suite 330, Boston,
MA 02111, USA.


ChangeLog
2005-04-21  Eric Smith <eric@brouhaha.com>

	* debugger.c: Only include vte.h when configured to use it, to
	avoid build problems.

2004-05-03  Eric Smith <eric@brouhaha.com>

	* csim.c (init_annunciator): Fix annunciator bitmap and GC creation.

	* csim.c (display_update, draw_annunciator): Draw annunciators.

2004-05-02  Eric Smith <eric@brouhaha.com>

	* csim.c (init_annunciator): Create bitmaps for annunciators.
	
	* kmly.y kml.c: Allow annunciator sections to be nested in display
	section.  Fix annunciator size parsing.  Print annunciators in
	print_kml().

2004-05-01  Eric Smith <eric@brouhaha.com>

	* coconut_lcd.c: Added coconut_op_display_compensation (acts as NOP).

	* voyager_lcd.c proc_nut.h proc_nut.c dis_nut.c: Added
	voyager_op_display_blink().

2004-04-30  Eric Smith <eric@brouhaha.com>

	* proc_woodstock.c dis_woodstock.h dis_woodstock.c: Added
	disassemblers for Woodstock and Nut processors.
	
	* proc_woodstock.c (woodstock_reset_processor): Only set external
	flag 5 (battery OK) on Woodstock series.

2004-04-29  Eric Smith <eric@brouhaha.com>

	* csim.c: Add --shape option and SHAPE_DEFAULT define.

2004-04-14  Eric Smith <eric@brouhaha.com>

	* proc_int.h proc.c proc_classic.c proc_woodstock.c: Remove glib.h
	dependency from proc_xxx.c.
	
	* kml.h kmll.l kmly.y csim.c (draw_digit): Segment definitions
	can be "line" or "rect".

2004-04-13  Eric Smith <eric@brouhaha.com>

	* csim.c display.h kml.c kml.h kmll.l kmly.y proc.c proc.h
	proc_int.h proc_classic.c proc_woodstock.c: Major rework of
	display code.

2004-04-12  Eric Smith <eric@brouhaha.com>

	* csim.c proc.h proc_int.h proc.c: Compute instruction rate from
	model and arch info.

	* arch.h arch.c: Replaced arch name table with struct arch_info_t
	and get_arch_info().

	* model.h model.c: Added clock frequency to model_info_t.

2004-04-11  Eric Smith <eric@brouhaha.com>

	* proc_woodstock.c (op_load_constant): Fix pointer wrap.

	* proc_woodstock.c (op_dec_p, op_inc_p): Increment and decrement P
	instructions wrap at word size, rather than setting P to an
	invalid position.

2004-04-10  Eric Smith <eric@brouhaha.com>

	* proc_woodstock.c (woodstock_handle_io): If the display is
	disabled, fill the string with spaces.
	
	* proc_woodstock.c (decode_woodstock_display): Add a trailing null
	character to the string.
	
2004-04-09  Eric Smith <eric@brouhaha.com>

	* proc_woodstock.c (woodstock_parse_object_line): Handle a
	five-digit octal address as a single-digit bank and four-digit
	address.

	* proc_woodstock.c (woodstock_print_state,
	woodstock_execute_instruction): Move call to
	woodstock_disassemble() into woodstock_print_state().

	* proc_woodstock.c (woodstock_print_state): Use env->prev_pc
	rather than env->pc, since PC has already been incremented.

	* proc_woodstock.c (op_arith, op_load_constant): When printing
	warning messages about P out of range, call
	woodstock_print_state().  Don't suppress message about loading an
	out of range P value, as the HP-45 is not a Woodstock.

	* proc_woodstock.c (woodstock_disassemble): Print opcode in octal
	rather than decimal.

	* proc_woodstock.c (woodstock_handle_io): Split display decode
	into multiple functions dispatched based on platform.  Implemented
	display decode for Spice platform.

	* csim.c (main) proc.c (sim_init) proc.h proc_int.h: Added
	platform field to sim_t.
	
	* model.c: Set RAM size for Spice models for which it is known,
	and ifdef'd out the models for which it isn't.
	
	* platform.h platform.c (find_platform_by_name): Added
	find_platform_by_name().

	* proc_woodstock.c (op_circulate_a_left, init_ops): Implemented
	0520 as CIRCULATE A LEFT instruction.  This is confirmed to be
	correct for the HP-33C (Spice), so it probably is for Woodstock as
	well.
	
2004-04-08  Eric Smith <eric@brouhaha.com>

	* proc_woodstock.c (op_clear_data_regs): The CLEAR DATA REGISTERS
	instruction only clears the single selected RAM chip (16
	registers).

	* proc.c proc.h proc_int.h proc_classic.h proc_woodstock.c csim.c:
	Added debug_flags field to sim_t, and sim_set_debug_flag() and
	sim_get_debug_flag() to control them.  Added debug menu commands
	to toggle debug flags.  All conditional on HAS_DEBUGGER.

	* proc_classic.c (classic_execute_instruction) proc_woodstock.c
	(woodstock_execute_instruction) proc_int.h: When tracing, call
	disassemble function if source code is not loaded.

	* proc_classic.c proc_woodstock.c (bad_op): Print opcode of
	illegal instructions in octal rather than hexadecimal.

2004-04-07  Eric Smith <eric@brouhaha.com>

	* all files: Merged from Woodstock branch.  This changed all files.
	The package has been renamed "nonpareil".

2004-04-04  Eric Smith <eric@brouhaha.com>

	* asm.h asm.c arch.h arch.c casmy.y wasmy.y csim.c Makefile: Added
	architecture and platform definitions.  Move model table from csim
	into arch.  Use architecture-specific listing file format.
	
	* asm.h: Added prototypes for yylex() and yyerror() to eliminate
	warnings.

	* asm.c (err_vprintf): Only output messages to listing file during
	pass 2.

	* asm.c (format_listing): Use return value from sprintf() to avoid
	unnecessary use of strlen().

	* wasml.l wasmy.y: Many fixes to Woodstock assembler; now HP-25
	source code assembles with no reported errors and object code
	matches listing.
	
	* asm.h asm.c casmy.y wasmy.y: Added a global symbol table for
	symbols starting with a '$'.
	
	* symtab.h symtab.c asm.h asm.c: Cleaned up typedefs, and use separate
	structs/typedefs for symbol table and symbol.

2004-04-03  Eric Smith <eric@brouhaha.com>

	* wasml.l wasmy.y Makefile: Started development of Woodstock
	assembler "wasm".

2004-04-02  Eric Smith <eric@brouhaha.com>

	* asm.h asm.c: Renamed from casm.h and casm.c.
	
	* casml.l casm.h casm.c: Moved keyword() and yywrap() into casm.c.

2004-03-24  Eric Smith <eric@brouhaha.com>

	* csim.c (main): Remove printf()s introduced to debug last change.

2004-03-23  Eric Smith <eric@brouhaha.com>


asm.c
/*
$Id$
Copyright 1995, 2004 Eric L. Smith <eric@brouhaha.com>

Nonpareil is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License version 2 as
published by the Free Software Foundation.  Note that I am not
granting permission to redistribute or modify Nonpareil under the
terms of any later version of the General Public License.

Nonpareil is distributed in the hope that it will be useful (or at
least amusing), but WITHOUT ANY WARRANTY; without even the implied
warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See
the GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program (in the file "COPYING"); if not, write to the
Free Software Foundation, Inc., 59 Temple Place - Suite 330, Boston,
MA 02111, USA.
*/

#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include "symtab.h"
#include "util.h"
#include "arch.h"
#include "asm.h"


int arch;


parser_t *parser [ARCH_MAX] =
  {
    [ARCH_UNKNOWN]   = asm_parse,
    [ARCH_CLASSIC]   = casm_parse,
    [ARCH_WOODSTOCK] = wasm_parse
  };


typedef void (write_obj_t)(FILE *f, int opcode);

static write_obj_t *write_obj [ARCH_MAX];


int pass;
int lineno;
int errors;
int warnings;

int group;	/* current rom group */
int rom;	/* current rom */
int pc;		/* current pc */

int dsr;	/* delayed select rom */
int dsg;	/* delayed select group */

char flag_char;

int objflag;	/* used to remember args to emit() */
int objcode;

int symtab_flag;

int last_instruction_type;


int targflag;	/* used to remember args to target() */
int targgroup;
int targrom;
int targpc;

char linebuf [MAX_LINE];
char *lineptr;

#define MAX_ERRBUF 2048
char errbuf [MAX_ERRBUF];
char *errptr;

#define SRC_TAB 32
char listbuf [MAX_LINE];
char *listptr;

#ifndef PATH_MAX
#define PATH_MAX 256
#endif

char srcfn  [PATH_MAX];
char objfn  [PATH_MAX];
char listfn [PATH_MAX];

FILE *srcfile  = NULL;
FILE *objfile  = NULL;
FILE *listfile = NULL;

symtab_t *global_symtab;
symtab_t *symtab [MAXGROUP] [MAXROM];  /* separate symbol tables for each ROM */


void usage (FILE *f)
{
  fprintf (f, "usage: %s objectfile\n", progname);
}


void format_listing (void)
{
  int i;

  listptr += sprintf (listptr, "%4d   ", lineno);

  if (arch == ARCH_CLASSIC)
    {
      if (objflag)
	{
	  listptr += sprintf (listptr, "L%1o%1o%03o:  ", group, rom, pc);
	  for (i = 0x200; i; i >>= 1)
	    *listptr++ = (objcode & i) ? '1' : '.';
	  *listptr = '\0';
	}
      else
	listptr += sprintf (listptr, "                   ");

      if (targflag)
	listptr += sprintf (listptr, "  -> L%1o%1o%03o", targgroup, targrom,
			    targpc);
      else
	listptr += sprintf (listptr, "           ");
	  
      listptr += sprintf (listptr, "  %c%c%c%c%c    ", flag_char, flag_char,
			  flag_char, flag_char, flag_char);
    }
  else
    {
      if (objflag)
	listptr += sprintf (listptr, "%06o  %04o ", objcode, (rom << 8) + pc);
      else
	listptr += sprintf (listptr, "             ");
    }
  
  strcat (listptr, linebuf);
  listptr += strlen (listptr);
}

void do_pass (int p)
{
  int i;

  arch = ARCH_UNKNOWN;
  
  pass = p;
  lineno = 0;
  errors = 0;
  warnings = 0;
  pc = 0;
  dsr = rom;
  dsg = group;
  last_instruction_type = OTHER_INST;

  printf ("Pass %d rom", pass);

  while (fgets (linebuf, MAX_LINE, srcfile))
    {
      /* remove newline */
      i = strlen (linebuf);
      if (linebuf [i - 1] == '\n')
	linebuf [i - 1] = '\0';

      lineno++;
      lineptr = & linebuf [0];

      listptr = & listbuf [0];
      listbuf [0] = '\0';

      errptr = & errbuf [0];
      errbuf [0] = '\0';

      objflag = 0;
      targflag = 0;
      flag_char = ' ';
      symtab_flag = 0;

      parser [arch] ();

      if (pass == 2)
	{
	  if (symtab_flag)
	    print_symbol_table (symtab [group] [rom], listfile);
	  else
	    {
	      format_listing ();
	      fprintf (listfile, "%s\n", listbuf);
	      if (errptr != & errbuf [0])
		{
		  fprintf (stderr, "%s\n", listbuf);
		  fprintf (listfile, "%s", errbuf);
		  fprintf (stderr, "%s",   errbuf);
		}
	    }
	}

      if (objflag)
	pc = (pc + 1) & 0xff;
    }

  if (pass == 2)
    {
      fprintf (listfile, "\nGlobal symbols:\n\n");
      print_symbol_table (global_symtab, listfile);
      fprintf (listfile, "\n");
    }

  printf ("\n");
}

void munge_filename (char *dst, char *src, char *ext)
{
  int i;
  int lastdot = 0;
  for (i = 0; src [i]; i++)
    {
      if (src [i] == '.')
	lastdot = i;
    }
  if (lastdot == 0)
    lastdot = strlen (src);
  memcpy (dst, src, lastdot);
  dst [lastdot] = '\0';
  strcat (dst, ext);
}

int main (int argc, char *argv[])
{
  progname = argv [0];

  if (argc != 2)
    {
      fprintf (stderr, "Usage: %s sourcefile\n", progname);
      exit (1);
    }

  global_symtab = alloc_symbol_table ();
  if (! global_symtab)
    fatal (2, "symbol table allocation failed\n");

  for (group = 0; group < MAXGROUP; group++)
    for (rom = 0; rom < MAXROM; rom++)
      {
	symtab [group] [rom] = alloc_symbol_table ();
	if (! symtab [group] [rom])
	  fatal (2, "symbol table allocation failed\n");
      }

  strcpy (srcfn, argv [1]);

  srcfile = fopen (srcfn, "r");

  if (! srcfile)
    fatal (2, "can't open input file '%s'\n", srcfn);

  munge_filename (objfn, srcfn, ".obj");
  munge_filename (listfn, srcfn, ".lst");

  objfile = fopen (objfn, "w");

  if (! objfile)
    fatal (2, "can't open input file '%s'\n", objfn);

  listfile = fopen (listfn, "w");

  if (! listfile)
    fatal (2, "can't open listing file '%s'\n", listfn);

  rom = 0;
  group = 0;

  do_pass (1);

  rewind (srcfile);

  do_pass (2);

  err_printf ("%d errors, %d warnings\n", errors, warnings);

  fclose (srcfile);
  fclose (objfile);
  fclose (listfile);
  exit (0);
}

void do_label (char *s)
{
  int prev_val;
  symtab_t *table;

  if (*s == '$')
    table = global_symtab;
  else
    table = symtab [group][rom];

  if (pass == 1)
    {
      if (! create_symbol (table, s, (rom << 8) + pc, lineno))
	error ("multiply defined symbol '%s'\n", s);
    }
  else if (! lookup_symbol (table, s, & prev_val))
    error ("undefined symbol '%s'\n", s);
  else if (prev_val != ((rom << 8) + pc))
    error ("phase error for symbol '%s'\n", s);
}


static void write_obj_classic (FILE *f, int opcode)
{
    fprintf (f, "%04o:%04o\n", (group << 11) | (rom << 8) | pc, opcode &01777);
}


static void write_obj_woodstock (FILE *f, int opcode)
{
    fprintf (f, "%04o:%04o\n", (rom << 8) | pc, opcode &01777);
}


static write_obj_t *write_obj [ARCH_MAX] =
  {
    [ARCH_CLASSIC]   = write_obj_classic,
    [ARCH_WOODSTOCK] = write_obj_woodstock
  };


static void emit_core (int op, int inst_type)
{
  objcode = op;
  objflag = 1;
  last_instruction_type = inst_type;

  if ((pass == 2) && objfile)
    write_obj [arch] (objfile, op);
}


void emit (int op)
{
  emit_core (op, OTHER_INST);
}

void emit_arith (int op)
{
  emit_core (op, ARITH_INST);
}

void emit_test (int op)
{
  emit_core (op, TEST_INST);
}

void target (int g, int r, int p)
{
  targflag = 1;
  targgroup = g;
  targrom = r;
  targpc = p;
}

int range (int val, int min, int max)
{
  if ((val < min) || (val > max))
    {
      error ("value out of range [%d to %d], using %d", min, max, min);
      return min;
    }
  return val;
}


/*
 * print to both listing file and standard error
 *
 * Use this for general messages.  Don't use this for warnings or errors
 * generated by a particular line of the source file.  Use error() or
 * warning() for that.
 */
int err_vprintf (char *format, va_list ap)
{
  int res;

  if (listfile && (pass == 2))
    vfprintf (listfile, format, ap);
  res = vfprintf (stderr, format, ap);
  return (res);
}

int err_printf (char *format, ...)
{
  int res;
  va_list ap;

  va_start (ap, format);
  res = err_vprintf (format, ap);
  va_end (ap);
  return (res);
}


/* generate error or warning messages and increment appropriate counter */
int error   (char *format, ...)
{
  int res;
  va_list ap;

  err_printf ("error in file %s line %d: ", srcfn, lineno);
  va_start (ap, format);
  res = err_vprintf (format, ap);
  va_end (ap);
  errptr += res;
  errors ++;
  return (res);
}

int warning (char *format, ...)
{
  int res;
  va_list ap;

  err_printf ("warning in file %s line %d: ", srcfn, lineno);
  va_start (ap, format);
  res = err_vprintf (format, ap);
  va_end (ap);
  errptr += res;
  warnings ++;
  return (res);
}


int keyword (char *string, keyword_t *table)
{
  while (table->name)
    {
      if (strcasecmp (string, table->name) == 0)
	return table->value;
      table++;
    }
  return 0;
}


